# Beginner Data Wrangling in Python
### Downloading StatsBomb Indian Super League 21/22 Data & digging into its structure

#### First, load the packages needed
#### Make sure to install any you don't have installed already

In [1]:
from mplsoccer import Sbopen
import pandas as pd

#### Then, initialize the StatsBomb parser so that we can get the data we want

In [2]:
parser = Sbopen(dataframe=True)

#### We download free StatsBomb data by passing a Match ID. This ID tells the code to download a specific match's data from the StatsBomb GitHub, essentially. Please see the 'ISL_21_22_MatchIDs' file on our GitHub to see all matches and their respective IDs.
#### We'll use the season opener between ATK Mohun Bagan and Kerala Blasters for this tutorial

In [3]:
match_id = 3813314

#### This line will download all of the events, shot freeze frames, starting lineups & formations, and other data for the given match


In [4]:
events, related, freeze, tactics = parser.event(match_id)

## Now all of the data for this one match is loaded as the dataframes 'events', 'related', etc. Let's take a look!

#### We'll mainly focus on the events here, but all of the data-wrangling concepts are the same for all datasets

In [5]:
events

id  index  period        timestamp  \
0     62ebdcd3-f4ad-4567-a17f-5c002ed1516f      1       1         00:00:00   
1     a4e2a845-6a89-4a0a-b3d1-65202310395d      2       1         00:00:00   
2     66461554-3613-476b-bf93-532b62a28320      3       1         00:00:00   
3     27baaeb2-92d8-4176-80d2-6826da141f88      4       1         00:00:00   
4     8f8aa851-bc54-4c6e-86c1-9a755fd245ac      5       1  00:00:01.125000   
...                                    ...    ...     ...              ...   
3022  e1651f42-5c92-49b2-9282-51b1f1324f0a   3023       2  00:50:36.576000   
3023  af312120-cacb-47f9-8b9f-5fb76d6b275f   3024       2  00:50:36.576000   
3024  1f5240d5-c63d-4079-a7d7-bc71cfee0692   3025       2  00:50:38.271000   
3025  dd958f35-f664-4221-9577-7d9c816a8191   3026       2  00:50:38.607000   
3026  b113221f-d4b2-4ad1-9f20-369a245b2450   3027       2  00:50:38.607000   

      minute  second  possession  duration  match_id  type_id  ...  \
0          0       0           1  0.000000   3813314       35  ...   
1          0       0           1  0.000000   3813314       35  ...   
2          0       0           1  0.000000   3813314       18  ...   
3          0       0           1  0.000000   3813314       18  ...   
4          0       1           2  1.875753   3813314       30  ...   
...      ...     ...         ...       ...       ...      ...  ...   
3022      95      36         197  0.000000   3813314        4  ...   
3023      95      36         197  0.000000   3813314        9  ...   
3024      95      38         197  0.000000   3813314        2  ...   
3025      95      38         197  0.000000   3813314       34  ...   
3026      95      38         197  0.000000   3813314       34  ...   

     ball_recovery_recovery_failure  dribble_nutmeg block_deflection  \
0                               NaN             NaN              NaN   
1                               NaN             NaN              NaN   
2                               NaN             NaN              NaN   
3                               NaN             NaN              NaN   
4                               NaN             NaN              NaN   
...                             ...             ...              ...   
3022                            NaN             NaN              NaN   
3023                            NaN             NaN              NaN   
3024                            NaN             NaN              NaN   
3025                            NaN             NaN              NaN   
3026                            NaN             NaN              NaN   

      pass_cut_back dribble_overrun  shot_deflected ball_recovery_offensive  \
0               NaN             NaN             NaN                     NaN   
1               NaN             NaN             NaN                     NaN   
2               NaN             NaN             NaN                     NaN   
3               NaN             NaN             NaN                     NaN   
4               NaN             NaN             NaN                     NaN   
...             ...             ...             ...                     ...   
3022            NaN             NaN             NaN                     NaN   
3023            NaN             NaN             NaN                     NaN   
3024            NaN             NaN             NaN                     NaN   
3025            NaN             NaN             NaN                     NaN   
3026            NaN             NaN             NaN                     NaN   

      bad_behaviour_card_id  bad_behaviour_card_name  pass_deflected  
0                       NaN                      NaN             NaN  
1                       NaN                      NaN             NaN  
2                       NaN                      NaN             NaN  
3                       NaN                      NaN             NaN  
4                       NaN                      NaN             NaN  
...                   

#### Lots of info there, but Jupyter doesn't show all the columns by default. So let's see all of them by running this next line

In [6]:
events.columns

Index(['id', 'index', 'period', 'timestamp', 'minute', 'second', 'possession',
       'duration', 'match_id', 'type_id', 'type_name', 'possession_team_id',
       'possession_team_name', 'play_pattern_id', 'play_pattern_name',
       'team_id', 'team_name', 'tactics_formation', 'off_camera', 'player_id',
       'player_name', 'position_id', 'position_name', 'pass_recipient_id',
       'pass_recipient_name', 'pass_length', 'pass_angle', 'pass_height_id',
       'pass_height_name', 'end_x', 'end_y', 'sub_type_id', 'sub_type_name',
       'body_part_id', 'body_part_name', 'x', 'y', 'under_pressure', 'out',
       'aerial_won', 'outcome_id', 'outcome_name', 'counterpress',
       'technique_id', 'technique_name', 'shot_statsbomb_xg',
       'shot_first_time', 'goalkeeper_position_id', 'goalkeeper_position_name',
       'pass_assisted_shot_id', 'pass_goal_assist', 'end_z',
       'shot_key_pass_id', 'pass_switch', 'pass_cross',
       'foul_committed_offensive', 'foul_won_defensive',
      

#### We can also just tell pandas to display all columns. This will allow you to see the first 5 and last 5 rows of the entire dataset. You can play around to see what it looks like

In [7]:
pd.set_option('display.max_columns', None)
events

id  index  period        timestamp  \
0     62ebdcd3-f4ad-4567-a17f-5c002ed1516f      1       1         00:00:00   
1     a4e2a845-6a89-4a0a-b3d1-65202310395d      2       1         00:00:00   
2     66461554-3613-476b-bf93-532b62a28320      3       1         00:00:00   
3     27baaeb2-92d8-4176-80d2-6826da141f88      4       1         00:00:00   
4     8f8aa851-bc54-4c6e-86c1-9a755fd245ac      5       1  00:00:01.125000   
...                                    ...    ...     ...              ...   
3022  e1651f42-5c92-49b2-9282-51b1f1324f0a   3023       2  00:50:36.576000   
3023  af312120-cacb-47f9-8b9f-5fb76d6b275f   3024       2  00:50:36.576000   
3024  1f5240d5-c63d-4079-a7d7-bc71cfee0692   3025       2  00:50:38.271000   
3025  dd958f35-f664-4221-9577-7d9c816a8191   3026       2  00:50:38.607000   
3026  b113221f-d4b2-4ad1-9f20-369a245b2450   3027       2  00:50:38.607000   

      minute  second  possession  duration  match_id  type_id      type_name  \
0          0       0           1  0.000000   3813314       35    Starting XI   
1          0       0           1  0.000000   3813314       35    Starting XI   
2          0       0           1  0.000000   3813314       18     Half Start   
3          0       0           1  0.000000   3813314       18     Half Start   
4          0       1           2  1.875753   3813314       30           Pass   
...      ...     ...         ...       ...       ...      ...            ...   
3022      95      36         197  0.000000   3813314        4           Duel   
3023      95      36         197  0.000000   3813314        9      Clearance   
3024      95      38         197  0.000000   3813314        2  Ball Recovery   
3025      95      38         197  0.000000   3813314       34       Half End   
3026      95      38         197  0.000000   3813314       34       Half End   

      possession_team_id possession_team_name  play_pattern_id  \
0                   7282      ATK Mohun Bagan                1   
1                   7282      ATK Mohun Bagan                1   
2                   7282      ATK Mohun Bagan                1   
3                   7282      ATK Mohun Bagan                1   
4                   7282      ATK Mohun Bagan                9   
...                  ...                  ...              ...   
3022                7283      Kerala Blasters                3   
3023                7283      Kerala Blasters                3   
3024                7283      Kerala Blasters                3   
3025                7283      Kerala Blasters                3   
3026                7283      Kerala Blasters                3   

     play_pattern_name  team_id        team_name  tactics_formation  \
0         Regular Play     7282  ATK Mohun Bagan             3412.0   
1         Regular Play     7283  Kerala Blasters              442.0   
2         Regular Play     7283  Kerala Blasters                NaN   
3         Regular Play     7282  ATK Mohun Bagan                NaN   
4        From Kick Off     7282  ATK Mohun Bagan                NaN   
...                ...      ...              ...                ...   
3022    From Free Kick     7283  Kerala Blasters                NaN   
3023    From Free Kick     7282  ATK Mohun Bagan                NaN   
3024    From Free Kick     7282  ATK Mohun Bagan                NaN   
3025    From Free Kick     7283  Kerala Blasters                NaN   
3026    From Free Kick     7282  ATK Mohun Bagan                NaN   

      off_camera  player_id                   player_name  position_id  \
0            NaN        NaN                           NaN          NaN   
1            NaN        NaN                           NaN          NaN   
2            NaN        NaN                           NaN          NaN   
3            NaN        NaN                           NaN          NaN   
4            1.0   163344.0                   Roy Krishna         22.0   
...          ...        ...                         

#### We see there are 3,027 total events in this data. These include everything from passes and shots to ball receipts and pressures, to the starting XIs and any formation changes throughout the match.
#### These events are all denoted in the 'type_name' column. So let's see all the events in this match

In [8]:
events.type_name.unique()

array(['Starting XI', 'Half Start', 'Pass', 'Ball Receipt', 'Duel',
       'Miscontrol', 'Carry', 'Pressure', 'Ball Recovery', 'Block',
       'Shield', 'Clearance', 'Shot', 'Goal Keeper', 'Dispossessed',
       'Dribble', 'Interception', 'Foul Committed', 'Foul Won',
       'Dribbled Past', 'Injury Stoppage', 'Substitution', 'Player Off',
       'Player On', 'Half End', 'Error', 'Referee Ball-Drop',
       'Tactical Shift', 'Bad Behaviour'], dtype=object)

#### There's another way to do this, by the way, and I'll include it so that you aren't getting errors with your own datasets (not this one) in the future.
#### If your column names have spaces or start with numbers, etc., you can't just use a period to call the column, such as events.column_name
#### You have to call the column name in those cases by putting it in [] and making it a string, such as events["column name"]

In [9]:
events['type_name'].unique()

array(['Starting XI', 'Half Start', 'Pass', 'Ball Receipt', 'Duel',
       'Miscontrol', 'Carry', 'Pressure', 'Ball Recovery', 'Block',
       'Shield', 'Clearance', 'Shot', 'Goal Keeper', 'Dispossessed',
       'Dribble', 'Interception', 'Foul Committed', 'Foul Won',
       'Dribbled Past', 'Injury Stoppage', 'Substitution', 'Player Off',
       'Player On', 'Half End', 'Error', 'Referee Ball-Drop',
       'Tactical Shift', 'Bad Behaviour'], dtype=object)

#### Now let's start manipulating & analyzing the data!
#### First, we'll filter to only show the events of Kerala Blasters' Magician, Adrián Luna

In [10]:
events[events['player_name']=='Adrián Nicolás Luna Retamar'].head()

id  index  period        timestamp  \
25  627701b2-efd6-4ebc-9499-0d71d62097ec     26       1  00:00:56.625000   
26  b7b8e203-b5b6-4c91-86de-6022e52f5701     27       1  00:00:56.625000   
27  b1462f75-810c-4f41-b781-2bdc45a7133d     28       1  00:00:56.871000   
48  cd7448ee-5c8d-4644-a122-b2f7a90f2f0e     49       1  00:01:58.376000   
53  2ee26eab-37b1-416c-a0d5-b4ac47e81af0     54       1  00:02:50.957000   

    minute  second  possession  duration  match_id  type_id     type_name  \
25       0      56           4       NaN   3813314       42  Ball Receipt   
26       0      56           4  0.245890   3813314       43         Carry   
27       0      56           4  2.723816   3813314       30          Pass   
48       1      58           7  0.239999   3813314       17      Pressure   
53       2      50           8       NaN   3813314       42  Ball Receipt   

    possession_team_id possession_team_name  play_pattern_id  \
25                7283      Kerala Blasters                4   
26                7283      Kerala Blasters                4   
27                7283      Kerala Blasters                4   
48                7282      ATK Mohun Bagan                2   
53                7283      Kerala Blasters                9   

   play_pattern_name  team_id        team_name  tactics_formation  off_camera  \
25     From Throw In     7283  Kerala Blasters                NaN         NaN   
26     From Throw In     7283  Kerala Blasters                NaN         NaN   
27     From Throw In     7283  Kerala Blasters                NaN         NaN   
48       From Corner     7283  Kerala Blasters                NaN         NaN   
53     From Kick Off     7283  Kerala Blasters                NaN         NaN   

    player_id                  player_name  position_id  \
25    26304.0  Adrián Nicolás Luna Retamar          9.0   
26    26304.0  Adrián Nicolás Luna Retamar          9.0   
27    26304.0  Adrián Nicolás Luna Retamar          9.0   
48    26304.0  Adrián Nicolás Luna Retamar          9.0   
53    26304.0  Adrián Nicolás Luna Retamar          9.0   

               position_name  pass_recipient_id    pass_recipient_name  \
25  Right Defensive Midfield                NaN                    NaN   
26  Right Defensive Midfield                NaN                    NaN   
27  Right Defensive Midfield            24035.0  Álvaro Vázquez García   
48  Right Defensive Midfield                NaN                    NaN   
53  Right Defensive Midfield                NaN                    NaN   

    pass_length  pass_angle  pass_height_id pass_height_name  end_x  end_y  \
25          NaN         NaN             NaN              NaN    NaN    NaN   
26          NaN         NaN             NaN              NaN   87.2   69.6   
27    21.545532   -1.505772             1.0      Ground Pass   88.6   48.1   
48          NaN         NaN             NaN              NaN    NaN    NaN   
53          NaN         NaN             NaN              NaN    NaN    NaN   

    sub_type_id sub_type_name  body_part_id body_part_name     x     y  \
25          NaN           NaN           NaN            NaN  87.3  69.3   
26          NaN           NaN           NaN            NaN  87.3  69.3   
27          NaN           NaN          40.0     Right Foot  87.2  69.6   
48          NaN           NaN           NaN            NaN  18.6  58.7   
53          NaN           NaN           NaN            NaN  45.3  37.0   

    under_pressure  out aerial_won  outcome_id outcome_name  counterpress  \
25             NaN  NaN        NaN         NaN          NaN           NaN   
26             NaN  NaN        NaN         NaN          NaN           NaN   
27             NaN  NaN        NaN         9.0   Incomplete           NaN   
48             NaN  NaN        NaN         NaN          NaN           NaN   
53             NaN  NaN        NaN         NaN          NaN           NaN   

    technique_id technique_name  shot_statsbomb_xg shot_first_time  \

#### That ".head()" at the end just shows the first 5 rows. You can change this to head(10) to show the first 10, and so on. ".tail()" is the opposite, and shows the last few rows
#### But wait, when we call our events dataframe again, it isn't filtered? Why?

In [11]:
events.head()

id  index  period        timestamp  \
0  62ebdcd3-f4ad-4567-a17f-5c002ed1516f      1       1         00:00:00   
1  a4e2a845-6a89-4a0a-b3d1-65202310395d      2       1         00:00:00   
2  66461554-3613-476b-bf93-532b62a28320      3       1         00:00:00   
3  27baaeb2-92d8-4176-80d2-6826da141f88      4       1         00:00:00   
4  8f8aa851-bc54-4c6e-86c1-9a755fd245ac      5       1  00:00:01.125000   

   minute  second  possession  duration  match_id  type_id    type_name  \
0       0       0           1  0.000000   3813314       35  Starting XI   
1       0       0           1  0.000000   3813314       35  Starting XI   
2       0       0           1  0.000000   3813314       18   Half Start   
3       0       0           1  0.000000   3813314       18   Half Start   
4       0       1           2  1.875753   3813314       30         Pass   

   possession_team_id possession_team_name  play_pattern_id play_pattern_name  \
0                7282      ATK Mohun Bagan                1      Regular Play   
1                7282      ATK Mohun Bagan                1      Regular Play   
2                7282      ATK Mohun Bagan                1      Regular Play   
3                7282      ATK Mohun Bagan                1      Regular Play   
4                7282      ATK Mohun Bagan                9     From Kick Off   

   team_id        team_name  tactics_formation  off_camera  player_id  \
0     7282  ATK Mohun Bagan             3412.0         NaN        NaN   
1     7283  Kerala Blasters              442.0         NaN        NaN   
2     7283  Kerala Blasters                NaN         NaN        NaN   
3     7282  ATK Mohun Bagan                NaN         NaN        NaN   
4     7282  ATK Mohun Bagan                NaN         1.0   163344.0   

   player_name  position_id         position_name  pass_recipient_id  \
0          NaN          NaN                   NaN                NaN   
1          NaN          NaN                   NaN                NaN   
2          NaN          NaN                   NaN                NaN   
3          NaN          NaN                   NaN                NaN   
4  Roy Krishna         22.0  Right Center Forward           124772.0   

  pass_recipient_name  pass_length  pass_angle  pass_height_id  \
0                 NaN          NaN         NaN             NaN   
1                 NaN          NaN         NaN             NaN   
2                 NaN          NaN         NaN             NaN   
3                 NaN          NaN         NaN             NaN   
4        Pritam Kotal     22.13504    2.568626             1.0   

  pass_height_name  end_x  end_y  sub_type_id sub_type_name  body_part_id  \
0              NaN    NaN    NaN          NaN           NaN           NaN   
1              NaN    NaN    NaN          NaN           NaN           NaN   
2              NaN    NaN    NaN          NaN           NaN           NaN   
3              NaN    NaN    NaN          NaN           NaN           NaN   
4      Ground Pass   41.4   52.0         65.0      Kick Off          38.0   

  body_part_name     x     y  under_pressure  out aerial_won  outcome_id  \
0            NaN   NaN   NaN             NaN  NaN        NaN         NaN   
1            NaN   NaN   NaN             NaN  NaN        NaN         NaN   
2            NaN   NaN   NaN             NaN  NaN        NaN         NaN   
3            NaN   NaN   NaN             NaN  NaN        NaN         NaN   
4      Left Foot  60.0  40.0             NaN  NaN        NaN         NaN   

  outcome_name  counterpress  technique_id technique_name  shot_statsbomb_xg  \
0          NaN           NaN           NaN            NaN                NaN   
1          NaN           NaN           NaN            NaN                NaN   
2          NaN           NaN           NaN            NaN                NaN   
3          NaN           NaN           NaN            NaN                NaN   
4          NaN           NaN           NaN            NaN  

#### This is because just filtering a dataset doesn't mean we've replaced the old one! So, let's create a new df so we have only Luna's events stored. We'll just name it luna. And make sure to remove 'head()', since we want ALL of Luna's events, not just the first 5!

In [12]:
luna = events[events['player_name']=='Adrián Nicolás Luna Retamar']
luna.head(3)

id  index  period        timestamp  \
25  627701b2-efd6-4ebc-9499-0d71d62097ec     26       1  00:00:56.625000   
26  b7b8e203-b5b6-4c91-86de-6022e52f5701     27       1  00:00:56.625000   
27  b1462f75-810c-4f41-b781-2bdc45a7133d     28       1  00:00:56.871000   

    minute  second  possession  duration  match_id  type_id     type_name  \
25       0      56           4       NaN   3813314       42  Ball Receipt   
26       0      56           4  0.245890   3813314       43         Carry   
27       0      56           4  2.723816   3813314       30          Pass   

    possession_team_id possession_team_name  play_pattern_id  \
25                7283      Kerala Blasters                4   
26                7283      Kerala Blasters                4   
27                7283      Kerala Blasters                4   

   play_pattern_name  team_id        team_name  tactics_formation  off_camera  \
25     From Throw In     7283  Kerala Blasters                NaN         NaN   
26     From Throw In     7283  Kerala Blasters                NaN         NaN   
27     From Throw In     7283  Kerala Blasters                NaN         NaN   

    player_id                  player_name  position_id  \
25    26304.0  Adrián Nicolás Luna Retamar          9.0   
26    26304.0  Adrián Nicolás Luna Retamar          9.0   
27    26304.0  Adrián Nicolás Luna Retamar          9.0   

               position_name  pass_recipient_id    pass_recipient_name  \
25  Right Defensive Midfield                NaN                    NaN   
26  Right Defensive Midfield                NaN                    NaN   
27  Right Defensive Midfield            24035.0  Álvaro Vázquez García   

    pass_length  pass_angle  pass_height_id pass_height_name  end_x  end_y  \
25          NaN         NaN             NaN              NaN    NaN    NaN   
26          NaN         NaN             NaN              NaN   87.2   69.6   
27    21.545532   -1.505772             1.0      Ground Pass   88.6   48.1   

    sub_type_id sub_type_name  body_part_id body_part_name     x     y  \
25          NaN           NaN           NaN            NaN  87.3  69.3   
26          NaN           NaN           NaN            NaN  87.3  69.3   
27          NaN           NaN          40.0     Right Foot  87.2  69.6   

    under_pressure  out aerial_won  outcome_id outcome_name  counterpress  \
25             NaN  NaN        NaN         NaN          NaN           NaN   
26             NaN  NaN        NaN         NaN          NaN           NaN   
27             NaN  NaN        NaN         9.0   Incomplete           NaN   

    technique_id technique_name  shot_statsbomb_xg shot_first_time  \
25           NaN            NaN                NaN             NaN   
26           NaN            NaN                NaN             NaN   
27           NaN            NaN                NaN             NaN   

    goalkeeper_position_id goalkeeper_position_name pass_assisted_shot_id  \
25                     NaN                      NaN                   NaN   
26                     NaN                      NaN                   NaN   
27                     NaN                      NaN                   NaN   

   pass_goal_assist  end_z shot_key_pass_id pass_switch pass_cross  \
25              NaN    NaN              NaN         NaN        NaN   
26              NaN    NaN              NaN         NaN        NaN   
27              NaN    NaN              NaN         NaN        NaN   

   foul_committed_offensive foul_won_defensive foul_committed_advantage  \
25                      NaN                NaN                      NaN   
26                      NaN                NaN                      NaN   
27                      NaN                NaN                      NaN   

   foul_won_advantage pass_shot_assist shot_one_on_one foul_committed_penalty  \
25                NaN              NaN             NaN                    NaN   
26                NaN              NaN             NaN 

#### Now pretend that we wanted to see just the passes Luna attempted. We'll need to filter again. And this time, let's overwrite our full Luna dataset instead of making a new one

In [13]:
luna = luna[luna['type_name']=='Pass']
luna.head(3)

id  index  period        timestamp  \
27   b1462f75-810c-4f41-b781-2bdc45a7133d     28       1  00:00:56.871000   
54   cfbf3880-09fa-4416-aaf4-78fd9ed975b0     55       1  00:02:52.082000   
120  6e78b864-6ec9-48d6-8d8e-59cbe73c2c31    121       1  00:03:46.469000   

     minute  second  possession  duration  match_id  type_id type_name  \
27        0      56           4  2.723816   3813314       30      Pass   
54        2      52           8  0.344657   3813314       30      Pass   
120       3      46          10  1.122585   3813314       30      Pass   

     possession_team_id possession_team_name  play_pattern_id  \
27                 7283      Kerala Blasters                4   
54                 7283      Kerala Blasters                9   
120                7283      Kerala Blasters                1   

    play_pattern_name  team_id        team_name  tactics_formation  \
27      From Throw In     7283  Kerala Blasters                NaN   
54      From Kick Off     7283  Kerala Blasters                NaN   
120      Regular Play     7283  Kerala Blasters                NaN   

     off_camera  player_id                  player_name  position_id  \
27          NaN    26304.0  Adrián Nicolás Luna Retamar          9.0   
54          NaN    26304.0  Adrián Nicolás Luna Retamar          9.0   
120         NaN    26304.0  Adrián Nicolás Luna Retamar          9.0   

                position_name  pass_recipient_id    pass_recipient_name  \
27   Right Defensive Midfield            24035.0  Álvaro Vázquez García   
54   Right Defensive Midfield           124750.0      Sahal Abdul Samad   
120  Right Defensive Midfield                NaN                    NaN   

     pass_length  pass_angle  pass_height_id pass_height_name  end_x  end_y  \
27     21.545532   -1.505772             1.0      Ground Pass   88.6   48.1   
54     48.349766    1.927873             2.0         Low Pass   53.9   65.7   
120    29.843927   -1.138623             3.0        High Pass  112.8   45.2   

     sub_type_id sub_type_name  body_part_id body_part_name      x     y  \
27           NaN           NaN          40.0     Right Foot   87.2  69.6   
54           NaN           NaN          40.0     Right Foot   70.8  20.4   
120          NaN           NaN          38.0      Left Foot  100.3  72.3   

     under_pressure  out aerial_won  outcome_id outcome_name  counterpress  \
27              NaN  NaN        NaN         9.0   Incomplete           NaN   
54              NaN  NaN        NaN         9.0   Incomplete           NaN   
120             1.0  NaN        NaN         9.0   Incomplete           NaN   

     technique_id technique_name  shot_statsbomb_xg shot_first_time  \
27            NaN            NaN                NaN             NaN   
54            NaN            NaN                NaN             NaN   
120           NaN            NaN                NaN             NaN   

     goalkeeper_position_id goalkeeper_position_name pass_assisted_shot_id  \
27                      NaN                      NaN                   NaN   
54                      NaN                      NaN                   NaN   
120                     NaN                      NaN                   NaN   

    pass_goal_assist  end_z shot_key_pass_id pass_switch pass_cross  \
27               NaN    NaN              NaN         NaN        NaN   
54               NaN    NaN              NaN        True        NaN   
120              NaN    NaN              NaN         NaN       True   

    foul_committed_offensive foul_won_defensive foul_committed_advantage  \
27                       NaN                NaN                      NaN   
54                       NaN                NaN                      NaN   
120                      NaN                NaN                      NaN   

    foul_won_advantage pass_shot_assist shot_one_on_one  \
27                 NaN              NaN             NaN   
54                 NaN              NaN             NaN   
1

#### After we have just our passes, why don't we see the average location Luna attempted passes from? We can use the '.mean()' function for this

In [14]:
print('Mean x:', luna.x.mean(), '\n', 'Mean y:', luna.y.mean())

Mean x: 79.4 
 Mean y: 57.040506329113924


#### StatsBomb use a pitch of 120 points in length (goal to goal) and 80 points in width. So we can see Luna's average pass attempt come just about at the start of the final third (average x, with x=80 being the start of the final third) and in the right half-space (average y)

#### Now let's get a little more complex to finish up this tutorial... Let's find out who Luna's main pass targets were this match. To do that, we have to perform something akin to a =sumif() or =countif() function in excel.
#### In Python, we can use the .groupby() operation

In [15]:
luna.groupby(['pass_recipient_name'])

#### Hey wait a minute where's our data!?

In [16]:
luna.groupby(['pass_recipient_name']).count()

id  index  period  timestamp  minute  second  \
pass_recipient_name                                                        
Ayush Adhikari               4      4       4          4       4       4   
Bijoy Varghese               5      5       5          5       5       5   
Chencho Gyeltshen            3      3       3          3       3       3   
Harmanjot Singh Khabra      10     10      10         10      10      10   
Jeakson Singh Thaunaojam     4      4       4          4       4       4   
Jessel Carneiro              2      2       2          2       2       2   
Jorge Rolando Pereyra Díaz   8      8       8          8       8       8   
Marko Lešković               8      8       8          8       8       8   
Prasanth Karuthadathkuni    11     11      11         11      11      11   
Rahul Kannoly Praveen        1      1       1          1       1       1   
Sahal Abdul Samad            8      8       8          8       8       8   
Soraisham Sandeep Singh      2      2       2          2       2       2   
Álvaro Vázquez García        4      4       4          4       4       4   

                            possession  duration  match_id  type_id  \
pass_recipient_name                                                   
Ayush Adhikari                       4         4         4        4   
Bijoy Varghese                       5         5         5        5   
Chencho Gyeltshen                    3         3         3        3   
Harmanjot Singh Khabra              10        10        10       10   
Jeakson Singh Thaunaojam             4         4         4        4   
Jessel Carneiro                      2         2         2        2   
Jorge Rolando Pereyra Díaz           8         8         8        8   
Marko Lešković                       8         8         8        8   
Prasanth Karuthadathkuni            11        11        11       11   
Rahul Kannoly Praveen                1         1         1        1   
Sahal Abdul Samad                    8         8         8        8   
Soraisham Sandeep Singh              2         2         2        2   
Álvaro Vázquez García                4         4         4        4   

                            type_name  possession_team_id  \
pass_recipient_name                                         
Ayush Adhikari                      4                   4   
Bijoy Varghese                      5                   5   
Chencho Gyeltshen                   3                   3   
Harmanjot Singh Khabra             10                  10   
Jeakson Singh Thaunaojam            4                   4   
Jessel Carneiro                     2                   2   
Jorge Rolando Pereyra Díaz          8                   8   
Marko Lešković                      8                   8   
Prasanth Karuthadathkuni           11                  11   
Rahul Kannoly Praveen               1                   1   
Sahal Abdul Samad                   8                   8   
Soraisham Sandeep Singh             2                   2   
Álvaro Vázquez García               4                   4   

                            possession_team_name  play_pattern_id  \
pass_recipient_name                                                 
Ayush Adhikari                                 4                4   
Bijoy Varghese                                 5                5   
Chencho Gyeltshen                              3                3   
Harmanjot Singh Khabra                        10               10   
Jeakson Singh Thaunaojam                       4                4   
Jessel Carneiro                                2                2   
Jorge Rolando Pereyra Díaz                     8                8   
Marko Lešković                                 8                8   
Prasanth Karuthadathkuni                      11               11   
Rahul Kannoly Praveen                          1                1   
Sahal Abdul Samad                              8                8   
Soraisha

#### Ah there we go. We need to add something to the end of the .groupby() operation. We have several options; .count(), .mean(), and .sum() are some main ones.
#### .count() will give us the number of instances (=countif() in excel)
#### .sum() will give us the sum of our variables (=sumif() in excel)
#### .mean() will give us the average of the variables (=averageif() in excel)

#### So with that, we know we want to use .count() to see the number of passes aimed at any specific player. nd as we saw above, just using .groupby().count() outputs all the data frame's numeric columns. With .count() they're all the same, so let's just show one column. For ease, the ID

In [17]:
luna.groupby(['pass_recipient_name'])['id'].count()

pass_recipient_name
Ayush Adhikari                 4
Bijoy Varghese                 5
Chencho Gyeltshen              3
Harmanjot Singh Khabra        10
Jeakson Singh Thaunaojam       4
Jessel Carneiro                2
Jorge Rolando Pereyra Díaz     8
Marko Lešković                 8
Prasanth Karuthadathkuni      11
Rahul Kannoly Praveen          1
Sahal Abdul Samad              8
Soraisham Sandeep Singh        2
Álvaro Vázquez García          4
Name: id, dtype: int64

#### But that doesn't look like our other data frame outputs. We need to do one more thing to make it look the same. We need to reset our data frame's index, so that the pass_recipient_name column is a column, instead of our data frame's index. The index is essentially what the rows are named. After resetting the index we see they're named 0, 1, 2 etc (in bold). I'm resetting here so that I can keep everything consistent and better show the next steps.

In [18]:
luna.groupby(['pass_recipient_name'])['id'].count().reset_index()

pass_recipient_name  id
0               Ayush Adhikari   4
1               Bijoy Varghese   5
2            Chencho Gyeltshen   3
3       Harmanjot Singh Khabra  10
4     Jeakson Singh Thaunaojam   4
5              Jessel Carneiro   2
6   Jorge Rolando Pereyra Díaz   8
7               Marko Lešković   8
8     Prasanth Karuthadathkuni  11
9        Rahul Kannoly Praveen   1
10           Sahal Abdul Samad   8
11     Soraisham Sandeep Singh   2
12       Álvaro Vázquez García   4

#### But now, the 'id' column is incorrectly named... that's not an ID but rather the number of passes to each player. So let's rename that

In [19]:
luna_passes = luna.groupby(['pass_recipient_name'])['id'].count().reset_index()

luna_passes.rename(columns={'id': 'passes'}, inplace=True)

luna_passes

pass_recipient_name  passes
0               Ayush Adhikari       4
1               Bijoy Varghese       5
2            Chencho Gyeltshen       3
3       Harmanjot Singh Khabra      10
4     Jeakson Singh Thaunaojam       4
5              Jessel Carneiro       2
6   Jorge Rolando Pereyra Díaz       8
7               Marko Lešković       8
8     Prasanth Karuthadathkuni      11
9        Rahul Kannoly Praveen       1
10           Sahal Abdul Samad       8
11     Soraisham Sandeep Singh       2
12       Álvaro Vázquez García       4

#### That 'inplace=True' at the end overwrites the luna_passes data frame for us, so we don't need to do luna_passes = luna_passes...
#### Now we're just about done, but the final step would be getting the most-passed-to players on top, so let's sort the data

In [20]:
luna_passes.sort_values(by=['passes'])

pass_recipient_name  passes
9        Rahul Kannoly Praveen       1
5              Jessel Carneiro       2
11     Soraisham Sandeep Singh       2
2            Chencho Gyeltshen       3
0               Ayush Adhikari       4
4     Jeakson Singh Thaunaojam       4
12       Álvaro Vázquez García       4
1               Bijoy Varghese       5
6   Jorge Rolando Pereyra Díaz       8
7               Marko Lešković       8
10           Sahal Abdul Samad       8
3       Harmanjot Singh Khabra      10
8     Prasanth Karuthadathkuni      11

#### Well... close! The .sort_values() operation defaults to lowest > highest. We need to add one last thing to change that

In [21]:
luna_passes.sort_values(by=['passes'], ascending=False)

pass_recipient_name  passes
8     Prasanth Karuthadathkuni      11
3       Harmanjot Singh Khabra      10
6   Jorge Rolando Pereyra Díaz       8
7               Marko Lešković       8
10           Sahal Abdul Samad       8
1               Bijoy Varghese       5
0               Ayush Adhikari       4
4     Jeakson Singh Thaunaojam       4
12       Álvaro Vázquez García       4
2            Chencho Gyeltshen       3
5              Jessel Carneiro       2
11     Soraisham Sandeep Singh       2
9        Rahul Kannoly Praveen       1

## And there we have it! Adrián Luna targeted RM Prasanth Karuthadathkuni 11 times, and RB Harmanjot Khabra 10 times.

#### Now you have the tools to download 21/22 ISL matches from StatsBomb's free data, and can begin manipulating it. Whether you're a pure beginner and all this was new, or a seasoned coder and just wanted to know how to download StatsBomb's data, I hope you learned something!
#### In future tutorials, we'll dive deeper into actual analysis and visualization of these match's data. Stay tuned!